src/workflow_provider/default_vm_workflow_provider.cpp

```diff
@@ -0,0 +1,376 @@
+#include <multipass/exceptions/download_exception.h>
+#include <multipass/exceptions/invalid_memory_size_exception.h>
+#include <multipass/exceptions/workflow_exceptions.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/poco_zip_utils.h>
+#include <multipass/query.h>
+#include <multipass/url_downloader.h>
+#include <multipass/utils.h>
+
+#include <QFile>
+#include <QFileInfo>
+
+#include <Poco/StreamCopier.h>
+#include <Poco/Zip/ZipStream.h>
+
+#include <fstream>
+#include <sstream>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+const QString github_workflows_archive_name{"multipass-workflows.zip"};
+const QString workflow_dir_version{"v1"};
+constexpr auto category = "workflow provider";
+
+auto workflows_map_for(const std::string& archive_file_path, bool& needs_update)
+{
+    std::map<std::string, YAML::Node> workflows_map;
+    std::ifstream zip_stream{archive_file_path, std::ios::binary};
+    auto zip_archive = MP_POCOZIPUTILS.zip_archive_for(zip_stream);
+
+    for (auto it = zip_archive.headerBegin(); it != zip_archive.headerEnd(); ++it)
+    {
+        if (it->second.isFile())
+        {
+            auto file_name = it->second.getFileName();
+            QFileInfo file_info{QString::fromStdString(file_name)};
+
+            if (file_info.dir().dirName() == workflow_dir_version &&
+                (file_info.suffix() == "yaml" || file_info.suffix() == "yml"))
+            {
+                if (!mp::utils::valid_hostname(file_info.baseName().toStdString()))
+                {
+                    mpl::log(
+                        mpl::Level::error, category,
+                        fmt::format("Invalid workflow name \'{}\': must be a valid host name", file_info.baseName()));
+                    needs_update = true;
+
+                    continue;
+                }
+
+                Poco::Zip::ZipInputStream zip_input_stream{zip_stream, it->second};
+                std::ostringstream out(std::ios::binary);
+                Poco::StreamCopier::copyStream(zip_input_stream, out);
+                workflows_map[file_info.baseName().toStdString()] = YAML::Load(out.str());
+            }
+        }
+    }
+
+    return workflows_map;
+}
+} // namespace
+
+mp::DefaultVMWorkflowProvider::DefaultVMWorkflowProvider(const QUrl& workflows_url, URLDownloader* downloader,
+                                                         const QDir& archive_dir,
+                                                         const std::chrono::milliseconds& workflows_ttl)
+    : workflows_url{workflows_url},
+      url_downloader{downloader},
+      archive_file_path{archive_dir.filePath(github_workflows_archive_name)},
+      workflows_ttl{workflows_ttl}
+{
+    try
+    {
+        update_workflows();
+    }
+    catch (const std::exception& e)
+    {
+        mpl::log(mpl::Level::error, category, fmt::format("Error on workflows start up: {}", e.what()));
+    }
```

Comment:
Since this is the ctor, if we don't catch everything, then it will cause daemon start up to bail and `multipassd` won't be running.  Is that what we want?

Reply:
We could avoid calling `update_workflows` in the c'tor then? We do have the `top_catch_all` bit to deal with at least some of that.

Reply:
I can certainly remove the preloading of workflows on daemon startup, but I don't understand why we would want to do that.  The way `DaemonConfig()` (which calls the `Workflow` ctor amongst many others) is currently architected, any exceptions are not caught and the daemon will bail.  I think the idea was to let the various classes determine if exceptions are catastrophic and let the daemon bail.  For `DefaultVMWorkflowProvider`, I don't see anything that should be catastrophic- the daemon should run if Workflows are unavailable for any reason which I why I have this in the ctor.

Reply:
The problem is, as I see it, you're left with a potentially broken workflow provider object. I'd be OK catching network, YAML parsing, ZIP errors, etc. Those we can recover from. But can we recover from _any_ exception?

If we caught "unexpected" exceptions in `DaemonConfig()`, we could NULL the provider and just not use it. But we shouldn't catch them all and ignore - and continue to use a potentially broken provider.

Reply:
Ok, I'll narrow the net to known exceptions, but I don't think this is the place to fix how `DaemonConfig()` doesn't handle exceptions that get bubbled up.

Reply:
I just removed the `try`/`catch` here and fixed the associated test.  `update_workflows()` catches the known exceptions, so any others will just go up to `DaemonConfig::build()`.